File: src/Library/Client/TransactedResponseProcessor.vb
Imports System.Threading

''' <summary>
''' Processes responses for transacted RCP methods.
''' </summary>
''' <remarks></remarks>
Friend Class TransactedResponseProcessor
    Inherits Pixa.Soundbridge.Library.ResponseProcessorBase

    Private _status As TransactionStatus = TransactionStatus.Pending

    ''' <summary>
    ''' Instantiates a new instance of ResponseProcessorBase for the specified SoundbridgeClient and EventWaitHandle.
    ''' </summary>
    ''' <param name="client"></param>
    ''' <param name="waitHandle"></param>
    ''' <remarks></remarks>
    Public Sub New(ByVal client As TcpSoundbridgeClient, ByVal command As String, ByVal waitHandle As EventWaitHandle)
        MyBase.New(client, command, waitHandle)
    End Sub

    ''' <summary>
    ''' Gets the status of the transaction.
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks>If the transaction was not initiated, this usually indicates an error.</remarks>
    Public ReadOnly Property Status() As TransactionStatus
        Get
            Return _status
        End Get
    End Property

    ''' <summary>
    ''' Processes the specified response line.
    ''' </summary>
    ''' <param name="response"></param>
    ''' <remarks></remarks>
    Public Overrides Sub Process(ByVal response As String)
        If response = "TransactionInitiated" Then
            _status = TransactionStatus.Initiated
            Exit Sub
        End If

        If response = "TransactionComplete" Then
            _status = TransactionStatus.Complete
            WaitHandle.Set()
            Exit Sub
        End If

        If response = "TransactionCanceled" Then
            _status = TransactionStatus.Canceled
            WaitHandle.Set()
            Exit Sub
        End If

        If ResponseCount = 0 Then
            If response = "StatusAwaitingReply" Then
                Client.OnAwaitingReply(Command)
                Exit Sub
            End If

            If response = "StatusSendingRequest" Then
                Client.OnSendingRequest(Command)
                Exit Sub
            End If

            If response.StartsWith("StatusReceivingData") Then
                If response.Contains(":") Then
                    response = response.Substring(21)

                    If response.Contains("/") Then
                        Dim parts As String() = response.Split("/")
                        Dim progress As Integer
                        Dim total As Integer

                        If Integer.TryParse(parts(0), progress) And Integer.TryParse(parts(1), total) Then
                            Client.OnReceivingData(Command, progress, total)
                        Else
                            Client.OnReceivingData(Command)
                        End If
                    Else
                        Dim progress As Integer

                        If Integer.TryParse(response, progress) Then
                            Client.OnReceivingData(Command, progress)
                        Else
                            Client.OnReceivingData(Command)
                        End If
                    End If
                Else
                    Client.OnReceivingData(Command)
                End If
            End If
        End If


        If response.StartsWith("ListResultSize") Or response = "ListResultEnd" Then Exit Sub

        AddResponse(response)

        If _status = TransactionStatus.Pending Then
            WaitHandle.Set()
            Exit Sub
        End If
    End Sub

    Public Overrides Sub PostProcess()
        If _status <> TransactionStatus.Complete And _status <> TransactionStatus.Canceled And ResponseCount = 0 Then ExceptionHelper.ThrowCommandTimeout(Command)
    End Sub
End Class

File: src/Pixa.Soundbridge/Client/TransactedResponseProcessor.cs
using System.Threading;

namespace Pixa.Soundbridge.Library
{

    /// <summary>
/// Processes responses for transacted RCP methods.
/// </summary>
/// <remarks></remarks>
    internal class TransactedResponseProcessor : ResponseProcessorBase
    {
        private TransactionStatus _status = TransactionStatus.Pending;

        /// <summary>
    /// Instantiates a new instance of ResponseProcessorBase for the specified SoundbridgeClient and EventWaitHandle.
    /// </summary>
    /// <param name="client"></param>
    /// <param name="waitHandle"></param>
    /// <remarks></remarks>
        public TransactedResponseProcessor(TcpSoundbridgeClient client, string command, EventWaitHandle waitHandle) : base(client, command, waitHandle)
        {
        }

        /// <summary>
    /// Gets the status of the transaction.
    /// </summary>
    /// <value></value>
    /// <returns></returns>
    /// <remarks>If the transaction was not initiated, this usually indicates an error.</remarks>
        public TransactionStatus Status
        {
            get
            {
                return _status;
            }
        }

        /// <summary>
    /// Processes the specified response line.
    /// </summary>
    /// <param name="response"></param>
    /// <remarks></remarks>
        public override void Process(string response)
        {
            if (response == "TransactionInitiated")
            {
                _status = TransactionStatus.Initiated;
                return;
            }

            if (response == "TransactionComplete")
            {
                _status = TransactionStatus.Complete;
                WaitHandle.Set();
                return;
            }

            if (response == "TransactionCanceled")
            {
                _status = TransactionStatus.Canceled;
                WaitHandle.Set();
                return;
            }

            if (ResponseCount == 0)
            {
                if (response == "StatusAwaitingReply")
                {
                    Client.OnAwaitingReply(Command);
                    return;
                }

                if (response == "StatusSendingRequest")
                {
                    Client.OnSendingRequest(Command);
                    return;
                }

                if (response.StartsWith("StatusReceivingData"))
                {
                    if (response.Contains(":"))
                    {
                        response = response.Substring(21);
                        if (response.Contains("/"))
                        {
                            var parts = response.Split('/');
                            int progress;
                            int total;
                            if (int.TryParse(parts[0], out progress) & int.TryParse(parts[1], out total))
                            {
                                Client.OnReceivingData(Command, progress, total);
                            }
                            else
                            {
                                Client.OnReceivingData(Command);
                            }
                        }
                        else
                        {
                            int progress;
                            if (int.TryParse(response, out progress))
                            {
                                Client.OnReceivingData(Command, progress);
                            }
                            else
                            {
                                Client.OnReceivingData(Command);
                            }
                        }
                    }
                    else
                    {
                        Client.OnReceivingData(Command);
                    }
                }
            }

            if (response.StartsWith("ListResultSize") | response == "ListResultEnd")
                return;
            AddResponse(response);
            if (_status == TransactionStatus.Pending)
            {
                WaitHandle.Set();
                return;
            }
        }

        public override void PostProcess()
        {
            if (_status != TransactionStatus.Complete & _status != TransactionStatus.Canceled & ResponseCount == 0)
                ExceptionHelper.ThrowCommandTimeout(Command);
        }
    }
}